core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -340,6 +351,12 @@ private void recoverFromTranslogInternal() throws IOException {
             flush(true, true);
         } else if (translog.isCurrent(translogGeneration) == false) {
             commitIndexWriter(indexWriter, translog, lastCommittedSegmentInfos.getUserData().get(Engine.SYNC_COMMIT_ID));
+            refreshLastCommittedSegmentInfos();
+        } else if (lastCommittedSegmentInfos.getUserData().containsKey(HISTORY_UUID_KEY) == false)  {
```

Comment:
do we also have to put it in the commit if the existing UUID is different to the one we have in the local var?

Reply:
with the current logic it will be a bug - the history should stay the same for the index life time and only go from null -> a value. With Yannick's suggestion it might be needed but I prefer to do it as a follow up.

Reply:
ok